### PR TITLE
[Perf] Disable clean_logits in deepgemm fp8_mqa_logits kernel

### PR DESCRIPTION
**Pull request by xyang16**
## Purpose

I noticed in DeepSeek V3.2 sparse_attn_indexer, `deep_gemm::smxx_clean_logits` kernel was launched following `deep_gemm::sm90_fp8_mqa_logits`, because `clean_logits` is set to true in https://github.com/vllm-project/vllm/blob/v0.16.0rc0/vllm/utils/deep_gemm.py#L331.

The purpose of `deep_gemm::smxx_clean_logits` kernel is to fill the padding value of MQA logits with -inf, see https://github.com/deepseek-ai/DeepGEMM/blob/v2.1.1/README.md?plain=1#L124. However, I found clean_logits is not necessary.

* For decode, the downstream `topKPerRowDecode` kernel read the row within the `rowEnd` range https://github.com/vllm-project/vllm/blob/v0.16.0rc0/csrc/sampler.cu#L575-L578, so it doesn't read padding values, therefore it is not necessary to enable clean_logits and fill padding values with -inf.
* Similarly for prefill, `topKPerRowPrefill` kernel doesn't read padding values https://github.com/vllm-project/vllm/blob/v0.16.0rc0/csrc/sampler.cu#L551-L553, so clean_logits is not necessary.
* Since `deep_gemm::smxx_clean_logits` kernel is not necessary and takes long time, eliminating launching this kernel would improve latency. So this PR disable clean_logits to improve performance.
* e2e output token throughput improves 15% on H200.

## Test Plan

Added test cases in `test_deepgemm_attention.py` and `test_top_k_per_row.py`

```
pytest -s -v tests/kernels/attention/test_deepgemm_attention.py
```

```
pytest -s -v tests/kernels/test_top_k_per_row.py
```

## Test Result

Unit tests passed.

## Profiling

* Prefill

Main:

543894508-14921f00-81ad-4859-97ed-fff5b092e1a2

PR:

543894111-b00e47e3-2a01-4494-9c86-22dcf453933c

* Decode

Main:

Screenshot 2026-02-01 at 10 20 33 PM

PR:

Screenshot 2026-02-01 at 9 21 15 PM

Main: `deep_gemm::smxx_clean_logits` kernel was launched.

PR: There's no `deep_gemm::smxx_clean_logits` kernel launched.

## Benchmark

Benchmarked on H200.

* DeepSeek V3.2

```
vllm serve deepseek-ai/DeepSeek-V3.2-Exp \
    --tensor-parallel-size 8 \
    --max-num-seqs 16 \
    --enable-expert-parallel \
    --trust-remote-code \
    --no-enable-prefix-caching
```
```
vllm bench serve \
  --model deepseek-ai/DeepSeek-V3.2-Exp \
  --dataset-name sharegpt \
  --dataset-path /tmp/ShareGPT_V3_unfiltered_cleaned_split.json \
  --sharegpt-output-len 300 \
  --max-concurrency 16 \
  --num-prompts 1000 \
  --num-warmups 50 \
  --ignore-eos
```

Main:

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Maximum request concurrency:             16        
Benchmark duration (s):                  661.38    
Total input tokens:                      229418    
Total generated tokens:                  300000    
Request throughput (req/s):              1.51      
Output token throughput (tok/s):         453.60    
Peak output token throughput (tok/s):    528.00    
Peak concurrent requests:                32.00     
Total token throughput (tok/s):          800.48    
---------------Time to First Token----------------
Mean TTFT (ms):                          1359.93   
Median TTFT (ms):                        477.05    
P99 TTFT (ms):                           11613.21  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          30.63     
Median TPOT (ms):                        30.46     
P99 TPOT (ms):                           31.99     
---------------Inter-token Latency----------------
Mean ITL (ms):                           30.63     
Median ITL (ms):                         30.47     
P99 ITL (ms):                            31.05     
==================================================

```

PR:

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Maximum request concurrency:             16        
Benchmark duration (s):                  576.87    
Total input tokens:                      229418    
Total generated tokens:                  300000    
Request throughput (req/s):              1.73      
Output token throughput (tok/s):         520.05    
Peak output token throughput (tok/s):    560.00    
Peak concurrent requests:                32.00     
Total token throughput (tok/s):          917.74    
---------------Time to First Token----------------
Mean TTFT (ms):                          433.33    
Median TTFT (ms):                        441.39    
P99 TTFT (ms):                           760.42    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          29.21     
Median TPOT (ms):                        29.19     
P99 TPOT (ms):                           30.44     
---------------Inter-token Latency----------------
Mean ITL (ms):                           29.21     
Median ITL (ms):                         29.20     
P99 ITL (ms):                            29.77     
==================================================
```

e2e output token throughput improves 15%.

* DeepSeek V3.2 MTP

```
vllm serve deepseek-ai/DeepSeek-V3.2-Exp \
    --tensor-parallel-size 8 \
    --max-num-seqs 16 \
    --enable-expert-parallel \
    --trust-remote-code \
    --no-enable-prefix-caching \
    --speculative-config '{"method": "mtp", "num_speculative_tokens": 1}'
```

Main:

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Maximum request concurrency:             16        
Benchmark duration (s):                  608.06    
Total input tokens:                      229418    
Total generated tokens:                  300000    
Request throughput (req/s):              1.64      
Output token throughput (tok/s):         493.37    
Peak output token throughput (tok/s):    432.00    
Peak concurrent requests:                22.00     
Total token throughput (tok/s):          870.66    
---------------Time to First Token----------------
Mean TTFT (ms):                          214.83    
Median TTFT (ms):                        241.03    
P99 TTFT (ms):                           387.68    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          31.57     
Median TPOT (ms):                        31.30     
P99 TPOT (ms):                           38.15     
---------------Inter-token Latency----------------
Mean ITL (ms):                           43.81     
Median ITL (ms):                         38.05     
P99 ITL (ms):                            171.86    
---------------Speculative Decoding---------------
Acceptance rate (%):                     38.90     
Acceptance length:                       1.39      
Drafts:                                  215488    
Draft tokens:                            215488    
Accepted tokens:                         83823     
Per-position acceptance (%):
  Position 0:                            38.90     
==================================================
```

PR:

```
============ Serving Benchmark Result ============
Successful requests:                     1000      
Failed requests:                         0         
Maximum request concurrency:             16        
Benchmark duration (s):                  588.06    
Total input tokens:                      229418    
Total generated tokens:                  300000    
Request throughput (req/s):              1.70      
Output token throughput (tok/s):         510.15    
Peak output token throughput (tok/s):    448.00    
Peak concurrent requests:                22.00     
Total token throughput (tok/s):          900.27    
---------------Time to First Token----------------
Mean TTFT (ms):                          210.88    
Median TTFT (ms):                        235.50    
P99 TTFT (ms):                           378.23    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          30.54     
Median TPOT (ms):                        30.34     
P99 TPOT (ms):                           37.12     
---------------Inter-token Latency----------------
Mean ITL (ms):                           42.37     
Median ITL (ms):                         36.72     
P99 ITL (ms):                            168.50    
---------------Speculative Decoding---------------
Acceptance rate (%):                     38.90     
Acceptance length:                       1.39      
Drafts:                                  215496    
Draft tokens:                            215496    
Accepted tokens:                         83830     
Per-position acceptance (%):
  Position 0:                            38.90     
==================================================
```

## Accuracy Testing

* DeepSeek V3.2

```
python3 -m lm_eval --model local-completions \
  --model_args model=deepseek-ai/DeepSeek-V3.2-Exp,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=16 \
  --tasks gsm8k
```

Main:

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9560|±  |0.0056|
|     |       |strict-match    |     5|exact_match|↑  |0.9538|±  |0.0058|
```

PR:

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9568|±  |0.0056|
|     |       |strict-match    |     5|exact_match|↑  |0.9560|±  |0.0056|
```

* DeepSeek V3.2 MTP

Main:

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9591|±  |0.0055|
|     |       |strict-match    |     5|exact_match|↑  |0.9583|±  |0.0055|
```

PR:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9591|±  |0.0055|
|     |       |strict-match    |     5|exact_match|↑  |0.9583|±  |0.0055|
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: tests/kernels/attention/test_deepgemm_attention.py [hunk 1]
```diff
@@ -135,13 +138,14 @@ def test_deepgemm_fp8_mqa_logits():
...
                     cu_seqlen_ke=ke,
                 )
-
                 ref_neginf_mask = ref_logits == float("-inf")
-                neginf_mask = logits == float("-inf")
-                assert torch.equal(neginf_mask, ref_neginf_mask)
+
+                if not clean_logits:
+                    neginf_mask = logits == float("-inf")
+                    assert torch.equal(neginf_mask, ref_neginf_mask)
```
[the remark this answers is not in the document]
Comment (xyang16) [reply tests/kernels/attention/test_deepgemm_attention.py]: Fixed.

**Review by yewentao256 — COMMENTED** [review]
LGTM, thanks for the work!
Also CC @LyricZhao

**Review by LopezCastroRoberto — COMMENTED** [review]
LGTM too, thanks!

**Review by yewentao256 — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
Thanks!
